### PR TITLE
[WIP] add differents properties types for identifiable elements

### LINKED ISSUE #1105: Network's properties should have different types
* **Do you want to request a *feature* or report a *bug*?**
Feature

* **What is the current behavior?**
Network's properties are always `String`.

* **What is the expected behavior?**
Network's properties should be able to have different types (`boolean`, `double`, `string`...)

* **Please tell us about your environment:**
  - PowSyBl Version: 3.2.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by jlabous**
Signed-off-by: Jérémy LABOUS <jlabous@silicom.fr>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
fix #1105 

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [x] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Comment by geofjamg** [comment]
Should we consider using Optional ( and equivalent for primitive types like OptionalInt) to get a property?

File: iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy [hunk 1]
```diff
@@ -45,18 +45,16 @@ class IdentifiableExtensionGroovyTest {
     void testStringProperty() {
         assertFalse(s.hasProperty())
         assertNull(s.greeting)
-        s.greeting = "hello"
+        s.setProperty("greeting", "hello")
```
Comment (tadam50) [inline iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy]: Revert changes and change IdentifiableExtension.groovy

File: iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy [hunk 2]
```diff
@@ -45,18 +45,16 @@ class IdentifiableExtensionGroovyTest {
...
     }
 
     @Test
     void testBooleanProperty() {
         assertFalse(s.hasProperty())
         assertNull(s.ok)
-        s.ok = true
-        assertEquals("true", s.getProperty("ok"))
-        assertEquals(true, s.ok)
+        s.setBooleanProperty("ok", true)
```
Comment (tadam50) [inline iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy]: Revert changes and change IdentifiableExtension.groovy

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 3]
```diff
@@ -14,18 +14,44 @@ import com.powsybl.iidm.network.Identifiable
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
  */
 class IdentifiableExtension {
-
     static Object propertyMissing(Identifiable self, String name) {
         // first check if an extension exist then a property
         Extension extension = self.getExtensionByName(name)
-        extension != null ? extension : self.properties[name]
+        if (extension != null) {
```
Comment (tadam50) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy]: Clean code and delete useless `;`

File: iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy (hunk 1 above)
Comment (tadam50) [reply iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy]: Missing line : `assertEquals("hello", s.getProperty("greeting"))`

File: iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy (hunk 2 above)
Comment (tadam50) [reply iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy]: Missing previous line adapted : `assertEquals(true, s.getBooleanProperty("ok"))`
Keep previous line as : `assertEquals(true, s.ok)` and remove `assertTrue(s.ok)`

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergedLine.java [hunk 4]
```diff
@@ -77,23 +79,121 @@ private static String buildName(final DanglingLine dl1, final DanglingLine dl2)
         return name;
     }
 
+    private static final String INCONSISTENCY_WARN_EMPTY_SIDE_1 = "Inconsistencies of property '{}' between both sides of merged line. Side 1 is empty, keeping side 2 value '{}'";
```
Comment (tadam50) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergedLine.java]: Cf comment in NetworkImpl

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 5]
```diff
@@ -146,7 +146,7 @@ public String setProperty(String key, String value) {
         } else {
             notifyElementReplaced(key, oldValue, val);
         }
-        return oldValue != null ? (String) oldValue.getValue() : null;
+        return oldValue != null && val.getKey().equals(oldValue.getKey()) ? (String) oldValue.getValue() : null;
```
Comment (tadam50) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java]: Rename `val` into `newValue` or more explicit
You may add a new method : isSameType ? (as isValueFound)

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 6]
```diff
@@ -158,7 +158,7 @@ public Integer setIntegerProperty(String key, Integer value) {
         } else {
             notifyElementReplaced(key, oldValue, val);
         }
-        return oldValue != null ? (Integer) oldValue.getValue() : null;
+        return oldValue != null && val.getKey().equals(oldValue.getKey()) ? (Integer) oldValue.getValue() : null;
```
Comment (tadam50) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java]: See previous review comment

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 7]
```diff
@@ -170,7 +170,7 @@ public Double setDoubleProperty(String key, Double value) {
         } else {
             notifyElementReplaced(key, oldValue, val);
         }
-        return oldValue != null ? (Double) oldValue.getValue() : null;
+        return oldValue != null && val.getKey().equals(oldValue.getKey()) ? (Double) oldValue.getValue() : null;
```
Comment (tadam50) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java]: See previous review comment

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 8]
```diff
@@ -182,7 +182,7 @@ public Boolean setBooleanProperty(String key, Boolean value) {
         } else {
             notifyElementReplaced(key, oldValue, val);
         }
-        return oldValue != null ? (Boolean) oldValue.getValue() : null;
+        return oldValue != null && val.getKey().equals(oldValue.getKey()) ? (Boolean) oldValue.getValue() : null;
```
Comment (tadam50) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java]: See previous review comment

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+public class Properties {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Properties.class);
+
+    private static final String INCONSISTENCY_WARN_EMPTY_SIDE_1 = "Inconsistencies of property '{}' between both sides of merged line. Side 1 is empty, keeping side 2 value '{}'";
+    private static final String INCONSISTENCY_WARN_EMPTY_SIDE_2 = "Inconsistencies of property '{}' between both sides of merged line. Side 2 is empty, keeping side 1 value '{}'";
+    private static final String INCONSISTENCY_ERROR_BOTH_SIDES = "Inconsistencies of property '{}' between both sides of merged line. '{}' on side 1 and '{}' on side 2. Removing the property of merged line";
+
+    public static class Property {
+        private final Pair<Type, Object> entry;
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java:22]: Don't use a Pair, but two attributes instead.
To avoid mistake, this class should be private and not exposed in the public API.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Property property = (Property) o;
+            return Objects.equals(entry, property.entry);
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java:61]: `return Objects.equals(entry, property.entry)` seems to be sufficient: the entry contains the type and the value, and `o` knows its own class

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+        @Override
+        public String toString() {
+            return "Property{" +
+                    "entry=" + entry +
+                    '}';
+        }
+    }
+
+    private final Map<String, Property> propertyList = new HashMap<>();
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java:77]: Rename propertyList because it's not a list... properties is better.

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 12]
```diff
@@ -220,41 +222,62 @@ public boolean hasProperty() {
...
+    public Properties.Type getPropertyType(String key) {
         return index.getNetworkStream()
-                .map(n -> n.getProperty(key))
-                .filter(Objects::nonNull)
-                .findFirst()
-                .orElse(null);
+            .map(n -> n.getPropertyType(key))
+            .filter(Objects::nonNull)
+            .findFirst()
+            .orElse(null);
```
Comment (mathbagu) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java:236]: Configure your IDE with 8 spaces to continue indentation and revert all the unnecessary changes due to your bad configuration. This remark is general for all this PR.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        Sets.difference(dl2Properties, commonProperties).forEach(prop -> setDanglingLineProperty(propertyList, dl2, prop));
+        commonProperties.forEach(prop -> {
+            if (dl1.getPropertyType(prop).equals(dl2.getPropertyType(prop))) {
+                mergeProperty(dl1, dl2, prop, dl1.getPropertyType(prop));
+            } else {
+                LOGGER.error("Inconsistencies of property type for '{}' between both sides of merged line. '{}' on side 1 and '{}' on side 2. Removing the property of merged line",
+                    prop, dl1.getPropertyType(prop), dl2.getPropertyType(prop));
+            }
+        });
+    }
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java:170]: I'm not sure these methods should be there: it's too specific. Why do not move it to the DanglingLine class.

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+    public <P> Optional<P> getProperty(String key, P defaultValue) {
+        Optional<P> val = getProperty(key);
+        return Optional.of(val.orElse(defaultValue));
+    }
+
+    public static boolean isSameType(Properties.Property oldValue, Properties.Property newValue) {
+        return newValue.getType() == oldValue.getType();
+    }
```
Comment (mathbagu) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/util/Properties.java:124]: Is it necessary to make this method public

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 15]
```diff
@@ -83,31 +79,55 @@ public boolean hasProperty(String key) {
...
+        return oldValue != null && Properties.isSameType(oldValue, newValue) ? (P) oldValue.getValue() : null;
+    }
+
+    private void notifyElementAdded(String key, Object newValue) {
+        getNetwork().getListeners().notifyElementAdded(this, () -> "properties[" + key + "]", newValue);
+    }
+
+    private void notifyElementReplaced(String key, Object oldValue, Object newValue) {
+        getNetwork().getListeners().notifyElementReplaced(this, () -> "properties[" + key + "]", oldValue, newValue);
     }
```
Comment (mathbagu) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java:116]: Why do you extract these methods: it seems they are used only once.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 16]
```diff
@@ -83,31 +79,55 @@ public boolean hasProperty(String key) {
...
     @Override
     public Set<String> getPropertyNames() {
-        return properties.keySet().stream().map(Object::toString).collect(Collectors.toSet());
+        return properties.keySet();
+    }
+
+    @Override
+    public Boolean removeProperty(String key) {
+        boolean hasProperty = hasProperty(key);
+        if (hasProperty) {
```
Comment (mathbagu) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java:126]: why do you introduce a local variable if you never use it?

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 17]
```diff
@@ -9,23 +9,58 @@ package com.powsybl.iidm.network.scripting
...
+            extension
+        } else {
+            switch(self.getPropertyType(name)) {
+                case Properties.Type.BOOLEAN:
+                    self.getProperty(name).orElse(null)
+                    break;
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy:27]: Remove unnecessary `;`

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 18]
```diff
@@ -9,23 +9,58 @@ package com.powsybl.iidm.network.scripting
...
+                case Properties.Type.DOUBLE:
+                    if (self.getProperty(name).isPresent()) {
+                        self.getProperty(name).get()
+                    }
+                    break;
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy:32]: Same here

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 19]
```diff
@@ -9,23 +9,58 @@ package com.powsybl.iidm.network.scripting
...
+                case Properties.Type.INTEGER:
+                    if (self.getProperty(name).isPresent()) {
+                        self.getProperty(name).get()
+                    }
+                    break;
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy:37]: Same here

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 20]
```diff
@@ -9,23 +9,58 @@ package com.powsybl.iidm.network.scripting
...
     static void propertyMissing(Identifiable self, String name, Object value) {
         if (value == null) {
-            self.properties.remove(name)
+            self.removeProperty(name)
         } else {
-            self.properties[name] = value;
+            switch(self.getPropertyType(name)) {
+                case Properties.Type.BOOLEAN:
+                    self.getProperty(name).orElse(null)
+                    break;
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy:50]: Same here

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 21]
```diff
@@ -9,23 +9,58 @@ package com.powsybl.iidm.network.scripting
...
+                case Properties.Type.DOUBLE:
+                    if (self.getProperty(name).isPresent()) {
+                        self.getProperty(name).get()
+                    }
+                    break;
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy:55]: Same here

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 22]
```diff
@@ -9,23 +9,58 @@ package com.powsybl.iidm.network.scripting
...
+                case Properties.Type.INTEGER:
+                    if (self.getProperty(name).isPresent()) {
+                        self.getProperty(name).get()
+                    }
+                    break;
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy:60]: Same here

File: iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy [hunk 23]
```diff
@@ -9,23 +9,58 @@ package com.powsybl.iidm.network.scripting
...
+                default:
+                    self.getProperty(name).orElse(null)
+            }
+        }
     }
-
     static void propertyMissing(Identifiable self, String name, Object value) {
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/main/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtension.groovy:43]: The implementation of this method is not correct: the property is created not get. Could you create a unit test that demonstrate the bug you make, and fix it?

File: iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy [hunk 24]
```diff
@@ -45,18 +45,18 @@ class IdentifiableExtensionGroovyTest {
...
     void testBooleanProperty() {
         assertFalse(s.hasProperty())
         assertNull(s.ok)
-        s.ok = true
-        assertEquals("true", s.getProperty("ok"))
+        s.setProperty("ok", true)
         assertEquals(true, s.ok)
+        assertEquals(true, s.getProperty("ok").get())
         s.ok = null
         assertNull(s.ok)
```
Comment (mathbagu) [inline iidm/iidm-scripting/src/test/groovy/com/powsybl/iidm/network/scripting/IdentifiableExtensionGroovyTest.groovy:61]: Do not change this unit test: this is the reason why you don't see your bug.

File: iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/AbstractNetworkTest.java [hunk 25]
```diff
@@ -151,42 +152,38 @@ public void testNetwork1() {
...
-        NetworkListener exceptionListener = mock(DefaultNetworkListener.class);
-        doThrow(new UnsupportedOperationException()).when(exceptionListener).onElementAdded(any(), anyString(), any());
-        doThrow(new UnsupportedOperationException()).when(exceptionListener).onElementReplaced(any(), anyString(),
-                any(), any());
         NetworkListener mockedListener = mock(DefaultNetworkListener.class);
 
         // Identifiable properties
         String key = "keyTest";
-        String value = "ValueTest";
+        String value = "valueTest";
```
Comment (mathbagu) [inline iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/AbstractNetworkTest.java:159]: Why do you change this?

**Comment by mathbagu** [comment]
The main remarks concern the API modification: I'm not sure we want to change the existing API. I think we should introduce generic get/set property overload, and consider that in the existing versions, the generic type is a String.

Looking to the test cases, it seems the API is not so fluent as we could expect, so we need to rework a little bit the PR. @MioRtia Any feedback?

**Comment by MioRtia** [comment]
> More of a general remark: I'm not really sure we want `getProperty` to return generic types... Why not use `getStringProperty`, `getDoubleProperty`, etc? It is more consistent with similar existing mechanisms (`ModuleConfig.getStringPropety`, `ParameterDefaultValueConfig.getStringValue`, etc...). `getProperty` would be deprecated and return by default `getStringProperty`.
> 
> What do you think @mathbagu?

Admittedly, it would make adding new property types heavier but is it such a big constraint?

File: iidm/iidm-xml-converter/src/main/java/com/powsybl/iidm/xml/PropertiesXml.java [hunk 26]
```diff
@@ -19,24 +22,81 @@
...
+                    identifiable.setIntegerProperty(name, Integer.parseInt(value));
+                    break;
+                case TYPE_DOUBLE:
+                    identifiable.setDoubleProperty(name, Double.parseDouble(value));
+                    break;
+                case TYPE_BOOLEAN:
+                    identifiable.setBooleanProperty(name, Boolean.parseBoolean(value));
+                    break;
+                default:
+                    throw new PowsyblException("");
```
Comment (MioRtia) [inline iidm/iidm-xml-converter/src/main/java/com/powsybl/iidm/xml/PropertiesXml.java:97]: Add a complete message: `"Property type " + type + " is not supported"`

**Comment by mathbagu** [comment]
@jlabous Could you please rebase your branch onto `evolution_xiidm/1.3`: we create the whole v1.3 dataset in another PR to be focus on your changes in this PR.

**Comment by MioRtia** [comment]
@jlabous Could you please correct the conflicts in this PR?

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java [hunk 27]
```diff
@@ -75,26 +116,201 @@ default Properties getProperties() {
...
+    default Set<String> getDoublePropertyNames() {
+        throw new UnsupportedOperationException("Not yet implemented");
+    }
+
+    /**
+     * Get boolean properties key values.
+     */
+    default Set<String> getBooleanPropertyNames() {
+        throw new UnsupportedOperationException("Not yet implemented");
+    }
```
Comment (MioRtia) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java:312]: Not sure this is useful (what do you think @mathbagu?)

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/Properties.java [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+import com.powsybl.iidm.network.Identifiable;
+import com.powsybl.iidm.network.Identifiable.PropertyType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+final class Properties {
```
Comment (MioRtia) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/Properties.java:21]: I don't agree with what you have done: it is really not efficient (and the way the maps are run through can create big performance issues...). I think you should do something like this:
```java
Map<String, Property> properties = new HashMap<>();
class Property {
    enum Type { STRING, BOOLEAN, INTEGER, DOUBLE };
    private final Type type;
    private final Object value;
    public Property(int value) { this.type = INTEGER; this.value = value; }
    public Property(String value) { this.type = STRING; this.value = value; }
    public int intValue() {
        if (type == INTEGER) {
            return (int) value;
        }
        throw ...;
    }
    public String stringValue() {
        if (type == STRING) {
             return (String) value;
        }
    }
...
}
```

File: iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/VariantManagerImplTest.java [hunk 29]
```diff
@@ -53,11 +50,6 @@ public String getId() {
             return Optional.empty();
         }
 
-        @Override
-        public String getNameOrId() {
```
Comment (MioRtia) [inline iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/VariantManagerImplTest.java:57]: Why did you delete this method?

**Comment by MioRtia** [comment]
This should be rebased on `evolution_xiidm/1.4` and the modifications will be on IIDM 1.4

**Comment by tadam50** [comment]
> This should be rebased on `evolution_xiidm/1.4` and the modifications will be on IIDM 1.4

Done

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java (hunk 27 above)
Comment (tadam50) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java]: Methods remoded

File: iidm/iidm-mergingview/pom.xml [hunk 30]
```diff
@@ -80,13 +80,13 @@
...
             <groupId>${project.groupId}</groupId>
-            <artifactId>powsybl-iidm-impl</artifactId>
+            <artifactId>powsybl-iidm-test</artifactId>
             <version>${project.version}</version>
             <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>${project.groupId}</groupId>
-            <artifactId>powsybl-iidm-test</artifactId>
+            <artifactId>powsybl-iidm-impl</artifactId>
```
Comment (tadam50) [inline iidm/iidm-mergingview/pom.xml:89]: Revert this change. Bad new order.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java [hunk 31]
```diff
@@ -152,31 +147,136 @@ public boolean hasProperty(String key) {
...
 
     @Override
-    public String getProperty(String key, String defaultValue) {
-        Object val = properties.getOrDefault(key, defaultValue);
-        return val != null ? val.toString() : null;
+    public PropertyType getPropertyType(String key) {
+        if (properties.containsKey(key)) {
+            return properties.get(key).propertyType();
+        }
+        return null;
```
Comment (MioRtia) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractIdentifiable.java:159]: Not sure if it should return null or throw an exception (what do you think @annetill ?) In any case, the behavior should be described in the javadoc of the `Identifiable` API

File: iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/VariantManagerImplTest.java [hunk 32]
```diff
@@ -81,23 +78,78 @@ public boolean hasProperty(String key) {
         }
 
         @Override
-        public String getProperty(String key) {
+        public Set<String> getPropertyNames() {
```
Comment (MioRtia) [inline iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/VariantManagerImplTest.java:81]: return collections.emptySet()

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java [hunk 33]
```diff
@@ -113,42 +130,169 @@ default String getNameOrId() {
...
+    default boolean setBooleanProperty(String key, boolean value) {
+        throw new UnsupportedOperationException("Not yet implemented");
+    }
 
     /**
      * Get properties key values.
      */
-    Set<String> getPropertyNames();
+    default Set<String> getPropertyNames() {
+        throw new UnsupportedOperationException("Not yet implemented");
```
Comment (MioRtia) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/Identifiable.java:294]: return collections.emptySet()

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/AbstractIdentifiableAdapter.java [hunk 34]
```diff
@@ -111,11 +174,6 @@ public void setFictitious(boolean fictitious) {
         return getDelegate().getExtensions();
     }
 
-    @Override
-    public String getImplementationName() {
```
Comment (MioRtia) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/AbstractIdentifiableAdapter.java:115]: Why is it deleted?

File: iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java [hunk 35]
```diff
@@ -808,11 +904,6 @@ public HvdcLine getHvdcLine(final String id) {
         return workingNetwork.getExtensions();
     }
 
-    @Override
-    public String getImplementationName() {
```
Comment (MioRtia) [inline iidm/iidm-mergingview/src/main/java/com/powsybl/iidm/mergingview/MergingView.java:812]: Why is it deleted?

File: iidm/iidm-xml-converter/src/main/java/com/powsybl/iidm/xml/ShuntXml.java [hunk 36]
```diff
@@ -187,7 +187,7 @@ protected void readElement(String id, ShuntCompensatorAdder adder, NetworkXmlRea
         if (regId[0] != null) {
             context.getEndTasks().add(() -> sc.setRegulatingTerminal(TerminalRefXml.readTerminalRef(sc.getNetwork(), regId[0], regSide[0])));
         }
-        properties.forEach(sc::setProperty);
+        properties.forEach(sc::setStringProperty);
```
Comment (MioRtia) [inline iidm/iidm-xml-converter/src/main/java/com/powsybl/iidm/xml/ShuntXml.java:190]: Set according to the type

**Comment by MioRtia** [comment]
Another remark: more code has been added since last update of this PR. Please check that `get/setProperty` is not used in new code.

**Comment by MioRtia** [comment]
A lot has changed since this PR and since it is not really high priority, I am closing it until it is and adding the tag `PR: unresolved`
